compose.yaml

```
@@ -0,0 +1,36 @@
+x-app-template: &app-template
+  image: rails_on_docker_2023-api-training-server:0.1.0
+  build: .
+  volumes:
+    - .:/app
+    - bundle:/usr/local/bundle
+  working_dir: /app
+
+services:
+  app:
+    <<: *app-template
+    command: bin/rails server --binding=0.0.0.0
+    ports:
+      - "${APP_PORT:-3000}:3000"
+    environment:
+      POSTGRES_HOST:
```

ここは env で指定するのではなく、compose.yml で固定しちゃっていいんじゃないかなあ。
```suggestion
      POSTGRES_HOST: db
```

---

まあ、コンテナを利用する際にホスト名を変更したいというシチュエーションはまずないから、ハードコードしてしまっていいのかな。

---

でも、いざ変更したいとなったときに、 `compose.yaml` に値を指定してあると、上書きするためには、優先度的にコマンドライン上(環境変数 or 引数)で直接指定するしかないみたいです。それをやるくらいなら、他の環境変数と同様に `.env` 経由で指定できるようにしておいたほうがラクなのでは、と思います。

https://docs.docker.com/compose/environment-variables/envvars-precedence/

---

docker compose を使っていて (db コンテナを起動するようにしておいて)、他の値に変更したいというシチュエーションがあまり思いつかないですね。
柔軟性があるように見えて、その実特に柔軟ではないというのはあまり意味がないのでは。
設定項目が増えることによって余計なトラブルを引き起こすだけでは?

プロジェクトの全員が Linux/UNIX であるとかミドルウェアに強いというわけではなく、プログラムは書けるが環境については得意ではないというメンバーは存在します。
(スキルは人それぞれ偏っているので、Rails は得意ではないが〜〜が得意というのは普通にありえる)

そうしたときに、使わない設定項目を増やすことはマイナスの価値を生むというのも知っておくと良さそうです。
具体的なメリットのあるシチュエーションがあるのであれば、追加するのには反対ではないのですが、いまのところ思いつかないので、うーん 🤔 という印象です。

※ まあ、-1 するほどではないかな、というところなので、これがいいという意見であれば引き下がるていどのものなので、このコメントで最後にします。

---

なるほど、これもある種の認知負荷の問題なんですかね。確かによく変更する設定とほとんど変更しない設定が並んでいると、人によっては混乱の元になりそうですね。

というか、今回のDBのホスト名の場合は、環境変数だけ変更しても、結局 `compose.yaml` に記載されているDBコンテナのラベル名も変更しないとアクセスできないから、なおさら意味がなかったですね。

というわけで修正します。

---

はい、そのとおりですね

* db コンテナの名前を変えるのであれば、compose.yml を書き換えるはず
    * このときに .env を書き換えるという話?
    * ファイルが分かれている価値はありそう?
* db コンテナ以外のデーターベースに接続したい
    * ほんとにそういうことやる?

ということで、柔軟性が上がっているとはいえないという認識でした。

ですから「ほとんど変更しない設定」ではなく「まったく変更しない設定」という理解でした。

---

ここまで書いて気づいたけど、データベースの ID, パスワードも変えたいというニーズはなさそうな気がしました。
セキュアにする必要がある場所ではないですし、compose.yml に固定値を書いてもよさそう

※ MySQL では `MYSQL_ALLOW_EMPTY_PASSWORD: 'yes'` という設定をして、root ユーザのパスワード空にするアプローチをよくやりますね

---

こうしておくと、 `.env` を書き換えるだけで複数のコンテナの環境変数設定を同時に変更できるので、環境構築時のように試行錯誤しているときは便利なんですよね。でも、それが完了した後だと、確かに変更したいという状況はなかなかないのかな。

なくてもいいならないほうがいいの精神でこちらも修正しておきます。